### ISSUE TITLE
[Q]: How to add pagination easilly

### ISSUE THREAD

**Issue by ParadiseFallen** [graphql-aspnet/graphql-aspnet#149]
Hi. im looking for smth like 
```
https://chillicream.com/docs/hotchocolate/v13/fetching-data/pagination
```
Is there any way to configure it like this?

**Comment by kevin-carroll** [comment]
Hey @ParadiseFallen ,

Great question! The feature you're quoting from hot chocolate is an implementation of the [Cursor Connections specification](https://relay.dev/graphql/connections.htm). This library does not yet support that spec. There is currently no built in, standardized way of doing pagination through the library.

Happy to add it to the list of features to implement for 2024 though. Thanks for the idea!

Cheers,
Kevin

**Comment by ParadiseFallen** [comment]
how i can impliment it myself?

**Comment by kevin-carroll** [comment]
If you're just looking to implement a custom pagination solution there are a number of ways to go about it. 

**Simple Solution:**
At a minimum you'd want to accept arguments to a field that track items such as `startAt` and `pageSize` to indicate to your server how many items you want (pageSize) and from which record you want to start retrieving from (startAt) and respond appropriately from your data source.  You could even encapsulate that into an input object to make it a bit more standardized within your project and a custom output object to emit data in a standardized way for your clients.  There are  a lot of options that would depend on your specific scenario and what trade offs you want to work with. 

```csharp
public class UsersController : GraphController
{
   public IEnumerable<User> GetUsers(int startAt, int pageSize)
  {
      // retrieve users from your data source starting at the given record and 
      // returning the number indicating by pageSize
  }

}
```

**Generalized Solution:**
Taking it a step further to a more general purpose solution, such as an implementation of the specification, would require creating and registering a server extension to extend the library. An extension of this nature would touch almost every aspect of execution from type templating, schema validation rules, introspection data all the way to a field resolution middleware component to intercept and interpret a new set of internal objects.  It would be a large undertaking. 

Other examples of large extensions are the subscriptions library and file upload extension in the main library. Neither of these are similar enough to what would be necessary to build the connections extension but its a place to start if you want to learn more about implementing your own detailed custom extensions.

Hope this helps,
Kevin

**Comment by ParadiseFallen** [comment]
@kevin-carroll i want generic method so it return IQueryable<Items>.

**Comment by kevin-carroll** [comment]
I'm not sure I follow you.  There isn't a valid scenario where you can return an `IQueryable` to the graphql engine that it would accept as anything other than an `IEnumerable`.   GraphQL deals with lists and objects, not unresolved projections. 

If you are referring to the hot chocolate implementation of its [custom query projections](https://chillicream.com/docs/hotchocolate/v13/fetching-data/projections), which expose an `IQueryable` to its internal plumbing for automatic filtering and paging; that's a very specific feature they've chosen to implement that is built on on their own internal projection framework.  Its a very nice QoL feature for those that want to take advantage of it within their ecosystem but its not specifically related to GraphQL as a query specification or a generalized paging strategy like the connections spec.  

You could implement something like that with this library if you wanted. The plumbing and hooks are there to allow you to do it but it would be an extremely advanced build that requires careful design considerations. 

I'm happy to answer any specific questions about how the library functions as you take on such an advanced project but I'm not in a position to propose how you might build such a solution at this time.

Kevin

**Comment by ParadiseFallen** [comment]
@kevin-carroll yeah. i want to write some custom endpoint invocator that can chech is it IQueryable and contains attribute like [EnablePagination] so it can apply pagination on top of already done endpoint
